File: backend/app/core/exchange_constants.py
"""
Indian Exchange Market Hours and Codes
======================================
Authoritative reference for all exchanges available on Zerodha KiteConnect.

Exchange codes used by Zerodha (these are the actual values in Trade.exchange):
  NSE   — NSE Equity/Cash segment
  BSE   — BSE Equity/Cash segment
  NFO   — NSE Futures & Options (index + stock derivatives)
  BFO   — BSE Futures & Options (SENSEX, BANKEX derivatives)
  CDS   — NSE Currency Derivatives
  BCD   — BSE Currency Derivatives
  MCX   — Multi Commodity Exchange (metals, energy, agri)
  MCXSX — Legacy MCX-SX (merged into BSE for currency; no longer active)

Market hours are IST (Asia/Kolkata, UTC+5:30), Monday–Friday.
These hours do NOT account for exchange-declared holidays or special sessions
(Muhurat trading, budget day short sessions, etc.).
For behavioral alert gating, this is accurate enough — holiday suppression
would require a live holiday calendar lookup and is not worth the complexity.

Sources:
  NSE/BSE equity/F&O: NSE circular NSE/MSD/52994; BSE circular 20230101-7
  MCX: MCX circular MCX/CS/178/2023 and MCX website
  Currency: SEBI circular CIR/MRD/DRMNP/26/2013

MCX note: Non-agri commodities (metals: Gold, Silver, Copper, Zinc, Lead,
  Nickel, Aluminium; energy: Crude Oil, Natural Gas) trade until 23:30 IST.
  Agri commodities (Castor Seed, Cardamom, Mentha Oil, CPO, Cottonseed,
  Kapas, Rubber, Pepper, Guar) close at 17:00 IST.
  We use 23:30 for ALL MCX (conservative — never wrongly suppresses a
  metals/energy trader working the legitimate evening session; an agri
  trader at 22:00 is an edge case we can tolerate).
"""

from datetime import datetime, time
from zoneinfo import ZoneInfo

IST = ZoneInfo("Asia/Kolkata")


# ---------------------------------------------------------------------------
# Exchange definitions
# ---------------------------------------------------------------------------
EXCHANGES: dict[str, dict] = {
    "NSE": {
        "name": "National Stock Exchange",
        "full_name": "NSE Equity Cash Segment",
        "description": "NSE equity / cash — stocks, ETFs, sovereign gold bonds",
        "segment": "Equity",
        "instruments": ["EQ", "BE", "SM"],   # BE = Book Entry (T+0); SM = SME IPO
        "open_time":   time(9, 15),
        "close_time":  time(15, 30),
        "pre_open":    time(9, 0),            # Call auction 09:00–09:15, no continuous executions
    },
    "BSE": {
        "name": "Bombay Stock Exchange",
        "full_name": "BSE Equity Cash Segment",
        "description": "BSE equity / cash — stocks, ETFs, mutual fund units",
        "segment": "Equity",
        "instruments": ["EQ", "BE", "SM"],
        "open_time":   time(9, 15),
        "close_time":  time(15, 30),
        "pre_open":    time(9, 0),
    },
    "NFO": {
        "name": "NSE Futures & Options",
        "full_name": "NSE Futures & Options Segment",
        "description": (
            "NSE derivatives — index futures/options (NIFTY 50, BANK NIFTY, "
            "FINNIFTY, MIDCPNIFTY, NIFTY IT) and single-stock F&O. "
            "Weekly and monthly expiry contracts."
        ),
        "segment": "F&O",
        "instruments": ["FUT", "CE", "PE"],
        "open_time":   time(9, 15),
        "close_time":  time(15, 30),
        "pre_open":    None,
        # Weekly expiry: every Thursday (index options); monthly: last Thursday of month
        # Expiry day trading continues until 15:30 — no special close time
    },
    "BFO": {
        "name": "BSE Futures & Options",
        "full_name": "BSE Futures & Options Segment",
        "description": (
            "BSE derivatives — SENSEX futures/options (BSE Sensex 30 index), "
            "BANKEX futures/options (BSE banking index). "
            "Weekly and monthly expiry contracts."
        ),
        "segment": "F&O",
        "instruments": ["FUT", "CE", "PE"],
        "open_time":   time(9, 15),
        "close_time":  time(15, 30),
        "pre_open":    None,
        # SENSEX weekly expiry: every Friday
        # BANKEX weekly expiry: every Monday
    },
    "CDS": {
        "name": "NSE Currency Derivatives",
        "full_name": "NSE Currency Derivatives Segment",
        "description": (
            "NSE currency F&O — USDINR, EURINR, GBPINR, JPYINR "
            "(futures and options). Extended session vs equity."
        ),
        "segment": "Currency",
        "instruments": ["FUT", "CE", "PE"],
        "open_time":   time(9, 0),
        "close_time":  time(17, 0),
        "pre_open":    None,
    },
    "BCD": {
        "name": "BSE Currency Derivatives",
        "full_name": "BSE Currency Derivatives Segment",
        "description": "BSE currency derivatives — parallel to CDS, lower liquidity",
        "segment": "Currency",
        "instruments": ["FUT", "CE", "PE"],
        "open_time":   time(9, 0),
        "close_time":  time(17, 0),
        "pre_open":    None,
    },
    "MCX": {
        "name": "Multi Commodity Exchange",
        "full_name": "MCX Commodity Derivatives",
        "description": (
            "MCX commodity futures and options. "
            "Non-agri metals (Gold, GoldM, GoldPetal, Silver, SilverM, SilverMic, "
            "Copper, Zinc, Lead, Nickel, Aluminium) and energy (CrudeOil, CrudeOilM, "
            "NaturalGas, NaturalGasMini) trade 09:00–23:30 IST. "
            "Agri commodities (Castorseed, Cardamom, MenthOil, CPO, "
            "CottonseedOilcake, Kapas, Rubber, Pepper, GuarSeed, GuarGum) "
            "trade 09:00–17:00 IST. "
            "Using 23:30 as effective close for ALL MCX in this codebase."
        ),
        "segment": "Commodity",
        "instruments": ["FUT", "CE", "PE"],
        "open_time":   time(9, 0),
        "close_time":  time(23, 30),
        "pre_open":    None,
        # NON-AGRI commodities traded on MCX (trade until 23:30 IST):
        #   Metals: GOLD, GOLDM, GOLDPETAL, SILVER, SILVERM, SILVERMIC,
        #           COPPER, ZINC, LEAD, NICKEL, ALUMINIUM
        #   Energy: CRUDEOIL, CRUDEOILM, NATURALGAS, NATURALGASM
        # AGRI commodities (technically close at 17:00 but using 23:30 as ceiling):
        #   CASTORSEED, CARDAMOM, MENTHAOIL, CPO, COTTONSEED,
        #   KAPAS, RUBBER, PEPPER, GUARSEED, GUARGUM
    },
    "MCXSX": {
        "name": "MCX-SX (legacy)",
        "full_name": "MCX Stock Exchange — Currency Segment (legacy)",
        "description": (
            "MCX-SX was a separate exchange for currency derivatives. "
            "Merged into BSE's currency segment; no longer active. "
            "Kept for backward-compatibility with historical trade data."
        ),
        "segment": "Currency",
        "instruments": ["FUT", "CE", "PE"],
        "open_time":   time(9, 0),
        "close_time":  time(17, 0),
        "pre_open":    None,
    },
}


# ---------------------------------------------------------------------------
# Market hours helper
# ---------------------------------------------------------------------------

def is_market_open(exchange: str, dt: datetime | None = None) -> bool:
    """
    Return True if the exchange is in its active trading session at `dt`.
    Defaults to now (IST). Weekends always return False.
    Does NOT account for exchange holidays or special sessions.

    Usage:
        from app.core.exchange_constants import is_market_open
        if is_market_open("NFO"):
            ...  # alert is actionable right now
    """
    if dt is None:
        dt = datetime.now(IST)
    elif dt.tzinfo is None:
        dt = dt.replace(tzinfo=IST)
    else:
        dt = dt.astimezone(IST)

    # Weekends: Saturday=5, Sunday=6
    if dt.weekday() >= 5:
        return False

    info = EXCHANGES.get(exchange.upper())
    if not info:
        # Unknown exchange — fall back to NSE hours as a safe default
        return time(9, 15) <= dt.time() <= time(15, 30)

    return info["open_time"] <= dt.time() <= info["close_time"]


def get_close_time(exchange: str) -> time:
    """Return the market close time (IST) for an exchange."""
    info = EXCHANGES.get(exchange.upper(), {})
    return info.get("close_time", time(15, 30))


def get_open_time(exchange: str) -> time:
    """Return the market open time (IST) for an exchange."""
    info = EXCHANGES.get(exchange.upper(), {})
    return info.get("open_time", time(9, 15))

File: src/lib/exchangeConstants.ts
/**
 * Indian Exchange Market Hours and Zerodha Symbol Parser
 * =======================================================
 * Authoritative reference matching backend/app/core/exchange_constants.py
 *
 * Exchange codes (exact values from Trade.exchange in Zerodha API):
 *   NSE   — NSE Equity/Cash segment (stocks, ETFs)
 *   BSE   — BSE Equity/Cash segment (stocks, ETFs)
 *   NFO   — NSE F&O: NIFTY, BANKNIFTY, FINNIFTY, MIDCPNIFTY, stock F&O
 *   BFO   — BSE F&O: SENSEX options/futures, BANKEX options/futures
 *   CDS   — NSE Currency Derivatives (USDINR, EURINR, GBPINR, JPYINR)
 *   BCD   — BSE Currency Derivatives
 *   MCX   — Multi Commodity Exchange (metals, energy, agri)
 *   MCXSX — Legacy MCX-SX (merged into BSE; historical data only)
 *
 * Zerodha tradingsymbol formats:
 *   Monthly options:  SYMBOL + YY + MMM + STRIKE + CE/PE  (e.g., NIFTY25MAR22000CE)
 *   Monthly futures:  SYMBOL + YY + MMM + FUT              (e.g., NIFTY25MARFUT)
 *   Weekly options:   SYMBOL + YY + M + DD + STRIKE + CE/PE (e.g., NIFTY2531922000CE)
 *     Weekly month codes: 1–9 = Jan–Sep, O = Oct, N = Nov, D = Dec
 *   Equity:           Plain ticker                          (e.g., RELIANCE, INFY)
 *   MCX futures:      SYMBOL + YY + MMM + FUT               (e.g., CRUDEOIL25MARFUT)
 */

// ---------------------------------------------------------------------------
// Exchange market hours (IST, Mon–Fri; holidays not accounted for)
// ---------------------------------------------------------------------------

export interface ExchangeInfo {
  name: string;
  segment: 'Equity' | 'F&O' | 'Currency' | 'Commodity';
  openHour: number;   // IST hour (24h)
  openMin: number;
  closeHour: number;
  closeMin: number;
}

export const EXCHANGES: Record<string, ExchangeInfo> = {
  NSE:   { name: 'NSE Equity',              segment: 'Equity',    openHour: 9,  openMin: 15, closeHour: 15, closeMin: 30 },
  BSE:   { name: 'BSE Equity',              segment: 'Equity',    openHour: 9,  openMin: 15, closeHour: 15, closeMin: 30 },
  NFO:   { name: 'NSE F&O',                 segment: 'F&O',       openHour: 9,  openMin: 15, closeHour: 15, closeMin: 30 },
  BFO:   { name: 'BSE F&O (SENSEX/BANKEX)', segment: 'F&O',       openHour: 9,  openMin: 15, closeHour: 15, closeMin: 30 },
  CDS:   { name: 'NSE Currency',            segment: 'Currency',  openHour: 9,  openMin: 0,  closeHour: 17, closeMin: 0  },
  BCD:   { name: 'BSE Currency',            segment: 'Currency',  openHour: 9,  openMin: 0,  closeHour: 17, closeMin: 0  },
  MCX:   { name: 'MCX Commodity',           segment: 'Commodity', openHour: 9,  openMin: 0,  closeHour: 23, closeMin: 30 },
  // MCX note: Non-agri metals & energy trade until 23:30; agri until 17:00.
  // Using 23:30 for all MCX — never wrongly suppresses a metals/energy alert.
  MCXSX: { name: 'MCX-SX (legacy)',         segment: 'Currency',  openHour: 9,  openMin: 0,  closeHour: 17, closeMin: 0  },
};

/**
 * Returns true if the given exchange is currently in its trading session (IST).
 * Weekends always return false. Holidays not accounted for.
 */
export function isMarketOpen(exchange: string): boolean {
  const info = EXCHANGES[exchange?.toUpperCase()];

  // Compute current IST time without importing a library
  const now = new Date();
  const istOffset = 5 * 60 + 30; // IST = UTC+5:30 in minutes
  const utcMinutes = now.getUTCHours() * 60 + now.getUTCMinutes();
  const istMinutes = (utcMinutes + istOffset) % (24 * 60);
  // Day of week in IST
  const utcMs = now.getTime();
  const istMs = utcMs + istOffset * 60 * 1000;
  const istDate = new Date(istMs);
  const dayOfWeek = istDate.getUTCDay(); // 0=Sun, 6=Sat

  if (dayOfWeek === 0 || dayOfWeek === 6) return false;

  if (!info) {
    // Unknown exchange — default to NSE hours
    return istMinutes >= 9 * 60 + 15 && istMinutes <= 15 * 60 + 30;
  }

  const openMinutes  = info.openHour  * 60 + info.openMin;
  const closeMinutes = info.closeHour * 60 + info.closeMin;
  return istMinutes >= openMinutes && istMinutes <= closeMinutes;
}


// ---------------------------------------------------------------------------
// Zerodha tradingsymbol parser
// ---------------------------------------------------------------------------

const MONTH_NAMES = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
const MONTH_3LETTER = ['JAN','FEB','MAR','APR','MAY','JUN','JUL','AUG','SEP','OCT','NOV','DEC'];

// Weekly option month codes: single char → 0-based month index
const WEEKLY_MONTH_CODE: Record<string, number> = {
  '1': 0, '2': 1, '3': 2, '4': 3, '5': 4, '6': 5,
  '7': 6, '8': 7, '9': 8, 'O': 9, 'N': 10, 'D': 11,
};

export interface ParsedSymbol {
  underlying: string;     // NIFTY, BANKNIFTY, SENSEX, CRUDEOIL, RELIANCE, etc.
  expiry?: string;        // "Mar 19"  (weekly) | "Mar '25" (monthly)
  strike?: number;        // 22000
  optionType?: 'CE' | 'PE';
  isFuture?: boolean;
  isWeekly?: boolean;
}

// Patterns ordered by specificity: monthly options first, then monthly futures, then weekly.
const MONTHLY_OPTION_RE = /^([A-Z&]+?)(\d{2})(JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|NOV|DEC)(\d+)(CE|PE)$/;
const MONTHLY_FUTURE_RE = /^([A-Z&]+?)(\d{2})(JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|NOV|DEC)FUT$/;
const WEEKLY_OPTION_RE  = /^([A-Z&]+?)(\d{2})([1-9ONDond])(\d{2})(\d+)(CE|PE)$/;

export function parseZerodhaSymbol(sym: string): ParsedSymbol {
  const s = sym.trim().toUpperCase();

  // 1. Monthly option: NIFTY25MAR22000CE, SENSEX25MAR74000PE, CRUDEOIL25MAR6000CE
  const mo = MONTHLY_OPTION_RE.exec(s);
  if (mo) {
    const [, underlying, , mon, strike, type] = mo;
    const monIdx = MONTH_3LETTER.indexOf(mon);
    const yy = mo[2];
    return {
      underlying,
      expiry: `${MONTH_NAMES[monIdx]} '${yy}`,
      strike: parseInt(strike, 10),
      optionType: type as 'CE' | 'PE',
      isFuture: false,
      isWeekly: false,
    };
  }

  // 2. Monthly future: NIFTY25MARFUT, CRUDEOIL25MARFUT, RELIANCE25MARFUT
  const mf = MONTHLY_FUTURE_RE.exec(s);
  if (mf) {
    const [, underlying, yy, mon] = mf;
    const monIdx = MONTH_3LETTER.indexOf(mon);
    return {
      underlying,
      expiry: `${MONTH_NAMES[monIdx]} '${yy}`,
      isFuture: true,
      isWeekly: false,
    };
  }

  // 3. Weekly option: NIFTY2531922000CE, SENSEX2631974800PE, BANKNIFTY2531954000CE
  const wo = WEEKLY_OPTION_RE.exec(s);
  if (wo) {
    const [, underlying, , mCode, dd, strike, type] = wo;
    const monIdx = WEEKLY_MONTH_CODE[mCode.toUpperCase()] ?? 0;
    return {
      underlying,
      expiry: `${MONTH_NAMES[monIdx]} ${parseInt(dd, 10)}`,
      strike: parseInt(strike, 10),
      optionType: type as 'CE' | 'PE',
      isFuture: false,
      isWeekly: true,
    };
  }

  // 4. Plain equity / unrecognized
  return { underlying: s };
}

/**
 * Format a Zerodha tradingsymbol for human-readable display.
 *
 * Returns:
 *   primary  — the main label (e.g., "SENSEX 74,800 PE" or "NIFTY FUT")
 *   secondary — expiry/date label (e.g., "Mar 19" or "Mar '25"), or undefined for equity
 *
 * Examples:
 *   SENSEX2631974800PE → { primary: "SENSEX 74,800 PE", secondary: "Mar 19" }
 *   NIFTY25MAR22000CE  → { primary: "NIFTY 22,000 CE",  secondary: "Mar '25" }
 *   NIFTY25MARFUT      → { primary: "NIFTY FUT",         secondary: "Mar '25" }
 *   RELIANCE           → { primary: "RELIANCE" }
 *   CRUDEOIL25MARFUT   → { primary: "CRUDEOIL FUT",      secondary: "Mar '25" }
 */
export function formatSymbol(sym: string): { primary: string; secondary?: string } {
  const parsed = parseZerodhaSymbol(sym);

  if (!parsed.expiry) {
    return { primary: parsed.underlying };
  }

  if (parsed.isFuture) {
    return {
      primary: `${parsed.underlying} FUT`,
      secondary: parsed.expiry,
    };
  }

  // Option: format strike in Indian locale (e.g., 74,800 not 74800)
  const strikeStr = parsed.strike != null
    ? parsed.strike.toLocaleString('en-IN')
    : '';

  return {
    primary: `${parsed.underlying} ${strikeStr} ${parsed.optionType ?? ''}`.trim(),
    secondary: parsed.expiry,
  };
}
